src/System.Data.Common/src/System/Data/Common/DbDataReaderExtensions.cs

```diff
@@ -2,22 +2,97 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections.Generic;
+
 namespace System.Data.Common
 {
     public static class DbDataReaderExtensions
     {
         public static System.Collections.ObjectModel.ReadOnlyCollection<DbColumn> GetColumnSchema(this DbDataReader reader)
         {
-            if (reader.CanGetColumnSchema())
+            if (reader is null)
+            {
+                throw new ArgumentNullException(nameof(reader));
+            }
+            if (reader is IDbColumnSchemaGenerator schemaGenerator)
             {
-                return ((IDbColumnSchemaGenerator)reader).GetColumnSchema();
+                return schemaGenerator.GetColumnSchema();
             }
-            throw new NotSupportedException();
+            return GetColumnSchemaCompatibility(reader);
         }
 
         public static bool CanGetColumnSchema(this DbDataReader reader)
         {
-            return reader is IDbColumnSchemaGenerator;
+            return true;
```

Comment:
Just noting: this means that `CanGetColumnSchema()` is now meaningless, as any reader either supports it either natively or via this shim. This makes sense - the only truly unsupported cases were in .NET Core pre-2.0, where DataTable and GetSchemaTable() simply did not exist and a shim wasn't possible (which is why this new API came into being in the first place).

Reply:
Good point. If we address https://github.com/dotnet/corefx/issues/36244 (in a different PR), we should only _move_ the GetColumnSchema to the System.Data namespace.